casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -422,67 +381,36 @@ object Validate {
-        // his previous block created (current Latest Message of sender)
-        // and new block `b` (potential new Latest Message of sender)
-        val newSendersBlock = b
-        val newLMs          = ProtoUtil.toLatestMessageHashes(newSendersBlock.justifications)
-        val curLMs          = ProtoUtil.toLatestMessageHashes(curSendersBlock.justifications)
-        // We let checkEquivocations handle when sender uses old self-justification
-        val newLMsNoSelf = newLMs.filterNot(_._1 == b.sender)
+    checkJustificationRegression(b).map { isValidOpt =>
+      if (isValidOpt.getOrElse(true)) BlockStatus.valid.asRight[BlockError]
+      else BlockStatus.justificationRegression.asLeft
+    }
 
-        // Check each Latest Message for regression (block seq num goes backwards)
-        newLMsNoSelf.toList.tailRecM {
-          // No more Latest Messages to check
-          case Nil =>
-            Applicative[F].pure(BlockStatus.valid.asRight.asRight)
-          // Check if sender of LatestMessage does justification regression
-          case newLM :: tail =>
-            val (sender, newJustificationHash) = newLM
-            val noSenderInCurLMs               = !curLMs.contains(sender)
-            if (noSenderInCurLMs) {
-              // If there is no justification to compare with - regression is not possible
-              Applicative[F].pure(Left(tail))
-            } else {
-              val curJustificationHash = curLMs(sender)
-              def logWarn(currentHash: BlockHash, regressiveHash: BlockHash): F[Unit] = {
-                val msg = s"block ${PrettyPrinter.buildString(regressiveHash)} by ${PrettyPrinter
-                  .buildString(sender)} has a lower sequence number than ${PrettyPrinter
-                  .buildString(currentHash)}."
-                Log[F].warn(ignore(b, msg))
-              }
-              // Compare and check for regression
-              val regressionDetected = for {
-                newJustification <- s.dag.lookupUnsafe(newJustificationHash)
-                curJustification <- s.dag.lookupUnsafe(curJustificationHash)
-                regression       = (!newJustification.invalid && newJustification.seqNum < curJustification.seqNum)
-                _                <- logWarn(curJustificationHash, newJustificationHash).whenA(regression)
-              } yield regression
-              // Exit tailRecM when regression detected, or continue to check remaining Latest Messages.
-              regressionDetected.ifM(
-                BlockStatus.justificationRegression.asLeft.asRight.pure,
-                Applicative[F].pure(Left(tail))
-              )
-            }
+  def checkJustificationRegression[F[_]: Monad: BlockDagStorage](
+      b: BlockMessage
+  ): F[Option[Boolean]] =
+    for {
+      msgMap <- BlockDagStorage[F].getRepresentation.map(_.dagMessageState.msgMap)
+//      justifications = b.justifications.map(_.latestBlockHash).map(msgMap)
```

Comment:
Leftover?

Reply:
It's related to below TODO comment. All justification should be present in `msgMap` to check justification regression. This code should be used to get justifications instead.